### ISSUE TITLE
Insane amount of connections

### ISSUE THREAD

**Issue by csimi** [aerospike/aerospike-client-nodejs#138]
We have 3 Aerospike nodes each having the following limits:
LimitNOFILE=250000
proto-fd-max is the same amount
The machines are from OVH: 128GB of RAM and the CPU is a single E5-1660.

On the Node.js servers that use the cache I have the following setting: 'maxConnsPerNode': 10000,
I'm connecting from 10 front-end servers with 8 cores each for a total of 80 threads.
Just simply getting out of our local network and being able to use OVH's servers required us to massively increase the ip_conntrack setting on our gateway.

So we've managed to work it out and everything was fine until we've doubled the amount of set requests, then everything fell apart.
We've started getting the following errors:
```
Jul 21 2016 15:23:47 GMT: WARNING (demarshal): (thr_demarshal.c:235) less than ten percent file handles remaining: 250000 max 248180 inuse
```
and requests started either giving back an error or timing out.

Is this an issue with only the Node.js driver or an inherent limitation of the pipeline-less protocol?
If you have any idea how we could scale our TPS please let me know.
I'm attaching a screenshot of our AMC shortly after restart.

Thanks for your time.

![AMC](http://i.imgur.com/gylwJaE.png)

**Comment by jhecking** [comment]
80 clients x 10.000 connections/node/client = max. 800.000 connections/node

Did you test the application at different levels for maxConnsPerNode?

One factor that's pretty important is the latency between the clients and the server. It's not quite clear from your description whether the clients are running on the same OVH network or are located somewhere else? The higher the latency the more connections you need to achieve the same TPS.

Pipelining can indeed help in that situation. In fact we just published a [detailed analysis](http://www.aerospike.com/blog/aerospike-3-8-supercharges-cross-datacenter-replication/) of the performance improvements using pipelined operations in cross-datacenter replication (XDR). This is on the roadmap for the Node.js client as well. But I cannot promise any delivery dates at this point.

**Comment by csimi** [comment]
The Aerospike servers are in a different network, we're using OVH servers temporarily until we get a larger cluster up and running in our usual datacenter. Latency is around 14ms.
The problem with having a lower amount of maximum connections is that then the requests seem to fail on the client side as I don't think there is any queueing involved (at least operations kept failing with the following error until we increased the limit on the client: Max node/event loop [...] async connections would be exceeded: 300).
Neither situation is optimal as it doesn't really matter if requests fail on the client or the server, they fail nonetheless.
We've raised the limits on the servers now so it'll work fine for a while.
Looking forward to having pipelining in the Node.js client, we'll be monitoring the developments.
